server/src-lib/Hasura/Events/Lib.hs

```diff
@@ -166,49 +168,86 @@ retryAfterHeader = "Retry-After"
+  events0 <- popEventsBatch
+  go events0 0 False
   where
-    -- FIXME proper backpressure. See: #3839 
-    pushEvents = forever $ do
-      let EventEngineCtx q _ _ fetchI = eectx
-      eventsOrError <- runExceptT $ Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite) fetchEvents
-      case eventsOrError of
-        Left err     -> L.unLogger logger $ EventInternalErr err
-        Right events -> atomically $ mapM_ (TQ.writeTQueue q) events
-      sleep fetchI
-
-    -- TODO this has all events race. How do we know this is correct? Document.
-    consumeEvents = forever $
-      -- ensure async exceptions from link only raised between iterations of forever block:
-      mask_ $ do
-        event <- atomically $ do
-          let EventEngineCtx q _ _ _ = eectx
-          TQ.readTQueue q
-        -- FIXME proper backpressure. See: #3839 
-        t <- async $ runReaderT (processEvent event) (logger, httpMgr, eectx)
-        -- Make sure any stray exceptions are at least logged via 'forkImmortal':
-        link t
-
-    -- NOTE: Blocks in tryWebhook if >= _eeCtxMaxEventThreads invocations active.
+    fetchBatchSize = 100
+    popEventsBatch = do
+      let run = runExceptT . Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite)
+      run (fetchEvents fetchBatchSize) >>= \case
+          Left err -> do 
+            L.unLogger logger $ EventInternalErr err
+            return []
+          Right events -> 
+            return events
+
+    go :: [Event] -> Int -> Bool -> IO void
+    go events !fullFetchCnt !alreadyWarned = do
+      -- process events ASAP until we've caught up; then we can sleep
+      when (null events) $ sleep _eeCtxFetchInterval
+
+      -- Prefetch next events payload while concurrently working through our current batch.
+      -- NOTE: we probably don't need to prefetch so early, but probably not
+      -- worth the effort for something more fine-tuned
+      eventsNext <- withAsync popEventsBatch $ \eventsNextA -> do
+        -- process approximately in order, minding HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE:
+        forM_ events $ \event -> 
+          mask_ $ do
+            atomically $ do  -- block until < HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE threads:
+              capacity <- readTVar _eeCtxEventThreadsCapacity
+              check $ capacity > 0
+              writeTVar _eeCtxEventThreadsCapacity (capacity - 1) 
+            let restoreCapacity = liftIO $ atomically $ 
+                  modifyTVar' _eeCtxEventThreadsCapacity (+ 1)
+            t <- async $ flip runReaderT (logger, httpMgr) $ 
+                    processEvent event `finally` restoreCapacity
+            link t
+
+        -- return when next batch ready; some 'processEvent' threads may be running.
+        wait eventsNextA
+
+      let lenEvents = length events
+      if | lenEvents == fetchBatchSize -> do
+             -- When we choose to warn is fairly arbitrary:
+             let clearlyBehind = fullFetchCnt >= 3
```

Comment:
a better name than `fullFetchCnt` would help since it doesn't correlate really well with `clearlyBehind`

Reply:
maybe I can resolve this with a better comment. The idea is this line should read "we are `clearlyBehind` in our processing when the last 3 fetches have been full (i.e. only limited by our `LIMIT` in the query)". Let me know if you have suggestions to improve aside from a comment like that

Reply:
I think it's allright. Maybe we can just do `fullFetchCount` but optional.